server/base/src/main/java/org/apache/accumulo/server/conf/store/PropCacheId.java

```diff
@@ -0,0 +1,278 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.server.conf.store;
+
+import static org.apache.accumulo.core.Constants.ZCONFIG;
+import static org.apache.accumulo.core.Constants.ZNAMESPACES;
+import static org.apache.accumulo.core.Constants.ZNAMESPACE_CONF;
+import static org.apache.accumulo.core.Constants.ZTABLES;
+import static org.apache.accumulo.core.Constants.ZTABLE_CONF;
+
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+import org.apache.accumulo.core.data.InstanceId;
+import org.apache.accumulo.core.data.NamespaceId;
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.fate.zookeeper.ZooUtil;
+import org.apache.accumulo.server.ServerContext;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+/**
+ * Provides a strongly-typed id for storing properties in ZooKeeper. The path in ZooKeeper is
+ * determined by the instance id and the type (system, namespace and table), with different root
+ * paths.
+ * <p>
+ * Provides utility methods from constructing different id based on type and methods to parse a
+ * ZooKeeper path and return a prop cache id.
+ */
+public class PropCacheId implements Comparable<PropCacheId> {
+
+  public static final String PROP_NODE_NAME = "encoded_props";
+
+  // indices for path.split();
+  public static final int TYPE_TOKEN_POSITION = 3;
+  public static final int IID_TOKEN_POSITION = 2;
+  public static final int ID_TOKEN_POSITION = 4;
+
+  // remove starting slash from constant.
+  public static final String TABLES_NODE_NAME = ZTABLES.substring(1);
+  public static final String NAMESPACE_NODE_NAME = ZNAMESPACES.substring(1);
+
+  private final String path;
+  private final IdType idType;
+  private final NamespaceId namespaceId;
+  private final TableId tableId;
+
+  private PropCacheId(final String path, final IdType idType, final NamespaceId namespaceId,
+      final TableId tableId) {
+    this.path = path;
+    this.idType = idType;
+    this.namespaceId = namespaceId;
+    this.tableId = tableId;
+  }
+
+  /**
+   * Instantiate a system prop cache id using the instance id from the context.
+   *
+   * @param context
+   *          the system context specifying the instance id
+   * @return a prop cache id for system properties,
+   */
+  public static PropCacheId forSystem(final ServerContext context) {
+    return forSystem(context.getInstanceID());
+  }
+
+  /**
+   * Instantiate a system prop cache id.
+   *
+   * @param instanceId
+   *          the instance id.
+   * @return a prop cache id for system properties,
+   */
+  public static PropCacheId forSystem(final InstanceId instanceId) {
+    return new PropCacheId(ZooUtil.getRoot(instanceId) + ZCONFIG + "/" + PROP_NODE_NAME,
```

Comment:
Maybe as a follow-on.  There is a lot of intermediate caching already - that with a new cache implementation and being able to get properties as a single call instead of reading all of the children, the original assumptions made with caching can be simplified or eliminated.

Rather than add more caching now, things will be easier to determine where and how things can be cached is that code is simplified.  The simplification was deferred as part of this PR in the hope that keeping as much of the original code / functionality would make review of this PR somewhat simpler.

Reply:
Does that imply there is another PR coming before 2.1.0 for simplification? Or was it deferred to a later release?

Reply:
I think he's just saying that this PR is already quite big, and that optimization doesn't need to be done as part of this. If it is needed it could be done later, but there's no current plans to do so.

Reply:
At this point I believe the sentiment is to get a release sooner.  It will depend on time-frames and other priorities.  There may be some easier things that might want to be done now, but there seems to be other portions that would need to wait for 3.0.

I don't intent to tackle it now - at least until these changes are merged and then it can be revisited and prioritized wrt a release.

Reply:
Follow-on issue create in https://github.com/apache/accumulo/issues/2634